0015-how-to-document-cnd-http-api.adoc

```diff
@@ -0,0 +1,207 @@
+
+=== How to visualise specification ===
+
+==== Options ====
+
+===== https://petstore.swagger.io/[SwaggerUI] =====
+* Extremely popular (over 15k stars on https://github.com/swagger-api/swagger-ui[GitHub])
+* RPC-style.
+* Looks like this:
+image::How_to_visualise_documentation/Swagger_UI_2019-08-29_18-16-23.png[scaledwidth=100%]
+
+* Hides JSON bodies.
+
+===== https://redocly.github.io/redoc/[ReDoc] =====
+* Popular (over 1k stars on GitHub).
+* Very https://github.com/Redocly/redoc/#deployment[simple] to set up.
+* Three-column style.
+* It looks very good:
+image::How_to_visualise_documentation/redoc-demo_2019-08-29_18-29-34.png[scaledwidth=100%]
+
+===== http://cheesestore.github.io/[Spectacle] =====
+* Quite popular (about 1k stars on GitHub).
+* Looks similar to ReDoc:
+image::How_to_visualise_documentation/screenshot_2019-08-29_18-46-24.jpg[scaledwidth=100%]
+
+==== Recommendation ====
+ReDoc because it is easy and I think it looks nice.
+
+=== What does the documentation include? ===
+
+==== HTTP API Specification. ====
+Models services.
+
+==== JSON Schema. ====
+Models data.
+
+=== How to host documentation for current `master` HEAD ===
+If we go for the recommended option of using custom-id:redoc[ReDoc], it's https://www.npmjs.com/package/redoc#tldr[very easy]:
+
+. Include API specification in repository with cnd.
+. Host a website on GitHub Pages (for example).
+. Use a HTML tag that links to the API specification (there's even a https://www.npmjs.com/package/redoc#usage-as-a-react-component[React component]).
+
+=== How to document extension link relation types? ===
+I would argue that this is covered in https://github.com/comit-network/comit-rs/issues/843[the original issue]. Just replace `human-protocol-spec` key with a link to a static HTML page hosted on GitHub Pages (for example) where the meaning of `human-protocol-spec` is described.
+
+=== Client-side validation ===
+
+==== Context ====
+* We already define custom-id:json-schema-integration[JSON Schemas] for some data objects returned from the API.
+* We have https://github.com/comit-network/comit-i/issues/44[discussed] doing input validation on comit-i.
+* We currently reproduce server-side validation on comit-i.
+
+==== Purpose ====
+* Improve user experience for users of clients of our API.
+* Simplify the job of developers.
+
+==== Proposal ====
+* Define JSON Schemas for all data objects that our API returns.
+* Offer a copy of these contracts to clients in a programmatically accessible format.
+* Include link to in response header:
+[source,html]
+----
+Link: <http://example.com/schemas/swap.schema.json#>; rel=”describedby”
+----
+* Use in comit-i to prove that it works.
+
+==== References ====
+* https://apisyouwonthate.com/blog/the-many-amazing-uses-of-json-schema-client-side-validation[Blog on client-side validation based on JSON Schema].
```

Comment:
Could you descrive a bit more OpenAPI vs JSON schema? I am not clear on the concepts.

Reply:
Hopefully clearer now.

Reply:
What commit made it clearer?

Reply:
[https://github.com/comit-network/spikes/pull/38/commits/e23ed624e1ce84896b1a23e3e0cf3faa61ab9122](This one).